src/Agent.Worker/Build/FileContainerServer.cs

```diff
@@ -0,0 +1,149 @@
+﻿using Microsoft.VisualStudio.Services.Common;
+using Microsoft.VisualStudio.Services.FileContainer.Client;
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.Services.Agent.Worker.Build
+{
+    public class FileContainerServer
+    {
+        private readonly ConcurrentQueue<string> _fileUploadQueue = new ConcurrentQueue<string>();
+        private readonly ConcurrentBag<Exception> _exceptionsDuringFileUpload = new ConcurrentBag<Exception>();
+        private CancellationTokenSource _uploadCancellationTokenSource;
+
+        private Uri _projectCollectionUrl;
+        private VssCredentials _credential;
+        private Guid _projectId;
+        private long _containerId;
+        private string _containerPath;
+
+        private int _filesUploaded = 0;
+        private string _sourceParentDirectory;
+
+        private FileContainerHttpClient FileContainerHttpClient { get; }
+
+        public FileContainerServer(
```

Comment:
I have the same thought about this class, i want to merge the logic for parallel upload into artifactcommand.cs and leave the FCServer as a thin wrap around REST Client. 
Will do that as a separate PR when i add L0 tests for ##vso Commands.